Fixed #33333 -- Fixed setUpTestData() crash with models.BinaryField on PostgreSQL.

felixxm opened the pull request:
This makes `models.BinaryField` pickleable on PostgreSQL.

Regression in 3cf80d3fcf7446afdde16a2be515c423f720e54d.

Thanks Adam Zimmerman for the report.

ticket-33333

felixxm commented on django/db/models/base.py:
```diff
@@ -572,6 +572,11 @@ def __getstate__(self):
         """Hook to allow choosing the attributes to pickle."""
         state = self.__dict__.copy()
         state['_state'] = copy.copy(state['_state'])
+        state['_memoryview_attrs'] = []
+        for k, v in list(state.items()):
+            if isinstance(v, memoryview):
+                state['_memoryview_attrs'].append((k, bytes(v)))
+                state.pop(k)
```
This is a bit hacky (but backward compatible), so an alternative patches and suggestions are welcome!

tjni replied:
I learned a bit from this about memoryviews. Thank you!

It might not be worth it, since it's premature optimization, but if you are in the mood, here are a few ideas:

1. Assign the copied bytes to a local variable list and only add them to the state if the list is non-empty. This saves a dictionary entry if no memoryviews exist, and it saves a hash lookup into the state for each memoryview.

2. Pop entries from the state after creating the local variable list. This would allow iterating over state's items without copying them into a list.

felixxm commented:
@AdamZ Does it work for you?

AdamZ commented:
Yep, it works! The codebase I work on is running Python 3.6/Django 3.2, so I couldn't test the patch directly due to the walrus operator. But when I rewrote the `__setstate__()` changes to be 3.6-compatible, it worked correctly.

I don't know if this patch is intended to be backported to Django 3.2, but if so that would need to be changed for the backported version.

felixxm commented:
> Yep, it works! The codebase I work on is running Python 3.6/Django 3.2, so I couldn't test the patch directly due to the walrus operator. But when I rewrote the `__setstate__()` changes to be 3.6-compatible, it worked correctly.

True, thanks for checking, this must be Python 3.5+ compatible, so I removed the walrus operator.

tjni commented on django/db/models/base.py:
```diff
@@ -572,6 +572,16 @@ def __getstate__(self):
+        # memoryview cannot be pickled, so cast it to bytes and store
+        # separately.
+        _memoryview_attrs = []
+        for attr, value in list(state.items()):
```
We should be able to iterate now though `state.items()` directly without making a copy because state is no longer being mutated in the body of the loop.

tjni reviewed (commented):
LGTM!

kezabelle commented on docs/releases/3.2.10.txt:
```diff
@@ -10,4 +10,5 @@ Django 3.2.10 fixes a security issue with severity "low" and several bugs in
 
-* ...
+* Fixed a regression in Django 3.2 that caused a crash of ``setUpTestData()``
+  with ``BinaryField`` on PostgreSQL (:ticket:`33333`).
```
Not knowing the details, the ticket specifically mentions `BinaryField`, but ... 
A bit of googling suggests that people have had issues with _pickling_ other fields which apparently contain memoryviews. I'd assume the fix handles that indiscriminately (if it's even an issue, given it's about using `deepcopy` rather than `pickle`), but I wonder if `BinaryField` needs calling out vs. built-in fields which are backed by a memoryview....

https://github.com/noripyt/django-cachalot/issues/115 (mentions `PolygonField`, `PointField`)
https://github.com/noripyt/django-cachalot/issues/125 (sort of suggests GIS in general, but I know nothing of GIS :))

felixxm replied:
I considered handling only `memoryviews` for `BinaryFields` but I don't think it's worth the extra complexity. I also tried `PointField` after you comment and couldn't reproduce the same issue. IMO it's nice when release notes are concrete, `BinaryField` on PostgreSQL is the only confirmed case so it should be good to mention it explicitly.

I added a small note that `BinaryField` is memoryview-backed in this case. Do you think this could be helpful?

kezabelle replied:
Oh I definitely wasn't suggesting _only_ handling `memoryviews` for `BinaryField`. Indiscriminate is good in this case precisely because it should handle any other exotic fields which might involve a `memoryview`. Apologies if that was confusing.

My question was really around whether the other fields might have been quietly suffering from the same issue, and if so specifying them too (even though the _ticket_ was only about `BinaryField`). If the others don't exhibit the same problem, well that's just good news isn't it? :)

I appreciate your adding the note (as you say, it's nice when release notes are concrete), but I'm torn about it - `memoryview` feels almost like an implementation detail, but at the same time, it's the cause of the issue. I think it's _probably fine_ either way, your preference.

charettes reviewed (approved):
It'd be great if we could solve the underlying issue for `BinaryField` on PostgreSQL on the main branch eventually but that's the best intermediate solution I can think of in terms of backportability.

Thanks for leading this effort @felixxm 🙇

felixxm commented:
Thanks y'all for reviewing :star:
